plugin/caffe/caffe_operator-inl.h

```diff
@@ -0,0 +1,331 @@
+/*!
+ * Copyright (c) 2016 by Contributors
+ * \file caffe_operator-inl.h
+ * \brief Caffe Operator
+ * \author Haoran Wang 
+*/
+#ifndef PLUGIN_CAFFE_CAFFE_OPERATOR_INL_H_
+#define PLUGIN_CAFFE_CAFFE_OPERATOR_INL_H_
+
+#include <caffe/layer.hpp>
+#include <caffe/proto/caffe.pb.h>
+#include <caffe/blob.hpp>
+
+#include <dmlc/logging.h>
+#include <dmlc/parameter.h>
+#include <mxnet/operator.h>
+
+#include <map>
+#include <vector>
+#include <string>
+#include <utility>
+#include <iostream>
+#include <exception>
+
+#include "../../src/operator/operator_common.h"
+
+#include "caffe_common.h"
+#include "caffe_operator_util.h"
+#include "caffe_stream.h"
+#include "caffe_fieldentry.h"
+#include "caffe_blob.h"
+
+namespace mxnet {
+namespace op {
+
+struct CaffeOperatorParam : public dmlc::Parameter<CaffeOperatorParam> {
+  caffe::LayerParameter prototxt;
+  int in_num, w_num, out_num;
+
+  DMLC_DECLARE_PARAMETER(CaffeOperatorParam) { DMLC_DECLARE_FIELD(prototxt).set_default("layer{}")
+    .describe("Caffe's layer parameter");
+    DMLC_DECLARE_FIELD(in_num).set_range(0, 100).set_default(1)
+    .describe("Operator input number");
+    DMLC_DECLARE_FIELD(out_num).set_range(0, 100).set_default(1)
+    .describe("Operator output number");
+  }
+};
+
+
+/**
+ * \brief this is the implementation of caffe operator in caffe.
+ * \tparam xpu the device that the op will be executed on.
+ */
+template<typename xpu, typename Dtype>
+class CaffeOperator : public Operator {
```

Comment:
Yes, they could be merged but I'm afraid then the readability might be a problem.

The major differences between two classes are:
* op has weights to process while loss doesn't
* loss needs to set grad_scale explicitly while op passes out_grad
* arguments of operator-property. op returns "data_i" while loss returns "data & label"

I feel the code logic and class definition would be clearer if written separately. What do you think?